functions/index.js

```diff
@@ -0,0 +1,125 @@
+// when a document is created on the guest list.
+exports.addWelcomeMessages = functions.firestore
+.document('rooms/{roomId}/guests/{guestId}').onCreate((snap, context) => {
+
+  const roomID = context.params.roomId;
+  const newValue = snap.data();
+  const userName = newValue.name;
+
+  // Saves the new welcome message into the database messages collection
+  admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
+    name: 'Lounge Bot',
+    profilePicUrl: '/images/LoungeLogo.png',
+    text: `${userName} joined the room! Welcome!`,
+    timestamp: admin.firestore.FieldValue.serverTimestamp(),
+  });
+  console.log('Welcome message written to database.');
+});
+
+// when a user is deleted off the guest list  or signs out
+// a leave message is inputted into the chat
+exports.addLeaveMessages = functions.firestore
+.document('rooms/{roomId}/guests/{guestId}').onDelete((snap, context) => {
+
+  const roomID = context.params.roomId;
+  const deletedValue = snap.data();
+  const userName = deletedValue.name;
+
+  admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
+    name: 'Lounge Bot',
+    profilePicUrl: '/images/LoungeLogo.png',
+    text: `${userName} left the room`,
+    timestamp: admin.firestore.FieldValue.serverTimestamp(),
+  });
+  console.log('Leave message written to database.');
+});
+
+// on update of the video playback data the chat is notified
+// when the video is paused and when the timestamp of the video is moved
+exports.updatePlayBack = functions.firestore
+.document('rooms/{roomId}/CurrentVideo/{PlaybackData}').onUpdate((change, context) => {
+
+  const roomID = context.params.roomId;
+  console.log('video change state');
+
+  // Get an object representing the document
+  const previousValue = change.before.data();
+  const changeValue = change.after.data();
+
+  // access a particular field
+  const isPlaying = changeValue.isPlaying;
+  const previousVideoState = previousValue.isPlaying;
+  const timestampNow = changeValue.timestamp;
+  const timestampBefore = previousValue.timestamp;
+
+  if (isPlaying == false && timestampNow !== 0) {
+    admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
+      name: 'Lounge Bot',
+      profilePicUrl: '/images/LoungeLogo.png',
+      text: `Video paused`,
+      timestamp: admin.firestore.FieldValue.serverTimestamp(),
+    });
+  }
+  // notifies user if video is playing after prevously being paused 
+  if (previousVideoState == false && timestampNow !== 0) {
+    admin.firestore().collection('rooms').doc(roomID).collection('messages').add({
+      name: 'Lounge Bot',
+      profilePicUrl: '/images/LoungeLogo.png',
+      text: `Video is playing`,
+      timestamp: admin.firestore.FieldValue.serverTimestamp(),
+    });
+  }
+  // only want to record instances when the timestamp changes outside the sync window
+  if (Math.abs(timestampNow - timestampBefore) > 5) {
```

Comment:
May want to use the SYNC_WINDOW constant in case it's ever changed in the code base.

Reply:
good point, ill change that

Reply:
sadly, trying to export variables from the front end to this server file causes errors within node since it can't understand some of the frontend language, so it may have to be changed manually as well